src/metadata/persistence/templates.clj

```diff
@@ -216,29 +251,59 @@
     (update-attribute user id attr)
     (insert-attribute user attr)))
 
+(defn- update-attr-enum-values
+  [attr-id {enum-values :values}]
+  (delete :attr_enum_values (where {:attribute_id attr-id}))
+  (dorun (map-indexed (partial add-attr-enum-value attr-id) enum-values)))
+
+(defn- update-nested-attribute
+  [user parent-id order attr]
+  (let [attr-id (insert-or-update-attribute user attr)]
+    (insert-nested-attr parent-id order attr-id)
+    (update-attr-enum-values attr-id attr)
+    (update-nested-attributes user attr-id (:attributes attr))))
+
+(defn- update-nested-attributes
+  [user parent-id children]
+  (delete :attr_attrs (where {:parent_id parent-id}))
+  (dorun (map-indexed (partial update-nested-attribute user parent-id) children)))
+
 (defn- update-template-attribute
-  [user template-id order {enum-values :values id :id :as attr}]
+  [user template-id order attr]
   (let [attr-id (insert-or-update-attribute user attr)]
     (insert-template-attr template-id order attr-id)
-    (delete :attr_enum_values (where {:attribute_id attr-id}))
-    (dorun (map-indexed (partial add-attr-enum-value attr-id) enum-values))))
+    (update-attr-enum-values attr-id attr)
+    (update-nested-attributes user attr-id (:attributes attr))))
 
 (defn- template-attr-subselect
   []
   (subselect [:template_attrs :ta]
              (where {:attributes.id :ta.attribute_id})))
 
+(defn- child-attr-subselect
+  []
+  (subselect [:attr_attrs :aa]
+             (where {:attributes.id :aa.child_id})))
+
 (defn- attr-synonym-subselect
   []
   (subselect [:attr_synonyms :s]
              (where {:attributes.id :s.synonym_id})))
 
-(defn- delete-orphan-attributes
+(defn- delete-current-orphan-attributes
   []
   (delete :attributes
           (where (and (not (exists (template-attr-subselect)))
+                      (not (exists (child-attr-subselect)))
                       (not (exists (attr-synonym-subselect)))))))
 
+(defn delete-orphan-attributes
+  "Deletes orphan attributes from the database. Deleting an attribute with children can produce more
+   orphans, so this function repeatedly runs the deletion until there are no orphans left. This function
+   should only be called from within a database transaction."
+  []
+  (take-while pos? (repeatedly delete-current-orphan-attributes)))
```

Comment:
Maybe a `delete cascade` in the database would delete nested attributes when their join-table entries are deleted, since we do not currently allow attributes to be reused: a nested attribute is only used by the parent, so if the parent is deleted, it should be safe to cascade-delete the children as well.

Anyway, I don't think we allow actual deletion in the UI, and the `permanent-delete` endpoint is only used by QA currently, so it's probably not worth anymore effort for now.

Reply:
Actually, now that I think about it more, this might not be possible with the join-table. This kind of `delete cascade` might only be possible if the `attributes` table had a `parent_id` instead of the `attr_attrs` join-table 🤔

Reply:
Yeah, I couldn't think of a clean way to do it either. The only option that came to mind was to keep track of which child attributes are being deleted and explicitly delete them and their descendants. That's certainly possible, but since the cruft will eventually be cleaned up, it didn't seem worth the added complexity.

Reply:
Agreed, it doesn't seem worth it.